OC-1999: use slf4j instead of log4j

spacemanaki opened the pull request:
As a library, Conduit should not be depending on a concrete logger, and
instead should depend on a logging interface (facade) like SLF4j

davidinjc commented:
lgtm

nnadirashvili commented:
Was there any manual integration tests done for the following:

1. DW8 app (nventory)
2. DW6 app (Baywatch)
3. Tomcat (Fulfillment)

Just to make sure we understand possible dependency conflicts with this change.

spacemanaki commented:
No, but there really should not be any dependency problems caused by this. I'm confident in saying that if there are problems, those should be fixed at the site because they indicate some other issue with whatever app that is using Conduit and depending on its dependency on a concrete logger. Ultimately, a library like Conduit should not be (should never be?) depending on a concrete logger like this.

nnadirashvili commented:
Yes conduit shouldn't be using log4j and using slf4j - no questions asked.

But I know shiro uses slf4j as well as hibernate (not sure which versions of slf4j... I though it would be nice to check how well this change will work with other dependencies)... especially with apps like baywatch (old DW version and old version of hibernate).

spacemanaki commented:
@nnadirashvili I really, really don't think this is going to be a problem. The artifact I'm adding here (org.slf4j:slf4j-api) has literally no other dependencies. And the entire point of SLF4J is to handle dependency issues -- I guarantee you that the authors are better at not breaking backwards compatibility than we are.

I've spent the past week working through dependency problems elsewhere and this fix addresses some of those issues. I could have papered over it but I'm trying to do the right thing by fixing it at the root. I think we just have to deal with problems this might cause if and when they come up. (to be clear, I believe the chance of this change causing serious dependency problems to be close to zero)

nnadirashvili commented:
Thanks for finding the root cause... again I think it's the right thing to do!

I'm more concerned about cross compatibility rather than backward compatibility. Wanted to know how the expected behavior was verified.

spacemanaki commented:
@davidinjc @carlo-rtr could one of you merge this please?

davidinjc commented:
@nnadirashvili You bring up an interesting point. As an organization we do not impose a specific third-party version in our projects which is often the culprit for dependency problems. Also we can't possibly predict all the different ways that the client libraries will be combined and how their dependencies will resolve at runtime, so it's hard to even pick the right version like you suggest. Admittedly this creates a bit of an untenable situation for us. Generally I think our practice has been to keep bumping the version forward and drag the older versions up when a conflict arises. But I think this practice merits more thinking and discussion.

Having said that, this appears to be a pretty safe change based on their FAQ -- http://www.slf4j.org/faq.html#compatibility. This is blocking Aki so I'm going to merge it.

nnadirashvili commented:
@spacemanaki @davidinjc  I'll merge this code, but at least can we consider making version of the slf4j a bit more dynamic so that slf4j version control could be standardized form the stand alone consumer applications and not from the counduit library itself. This can help us avoid cross library dependency clashes.

Something similar to this:
    <properties>
        <powermock.version>1.4.11</powermock.version>
        <org.slf4j.version>1.6.4</org.slf4j.version> // Of course if 1.6.4 sits well with conduit it's consumers
    </properties>

        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-api</artifactId>
            <version>${org.slf4j.version}</version>
        </dependency>

This is a standard way of handling dependencies like this from third party libraries. I understand that this is a blocker, but proposed change is not big of a change either and could be done in 30 minutes if it makes sense of course.

On top of that, when it comes to changing code and there is a question of "how did you verify expected behavior" I expected a little more specific answers like I did "A", "B" & "C"... not like "I guarantee you this will work"... I'm pretty certain if I submitted a PR like this I'll have to answer tons of questions and my responses will be pretty detailed as to what steps I took to verify expected behavior (be it automated test or manual).

I think I raised enough concerns for this PR. If the majority feels that this is the way to go with our "library" projects, then feel free to merge it.

spacemanaki commented:
@nnadirashvili Thanks for your feedback! I'm going to try to start a much larger discussion of dependency handling in our Java services, and hopefully figure out some standards and guidelines for us to avoid these problems. I'm not sure if your suggestion will be enough to address some of the problems I ran into, but you're obviously welcome to participate in that discussion when it happens, and I hope you do! We will need all the help we can get...

Your example of auth is interesting -- on the one hand that library looks like it depends on two concrete logging packages with `<scope>runtime</scope>` which should not result in those being transitive dependencies in an app that uses auth: https://github.com/RentTheRunway/auth/blob/master/client/pom.xml#L90-L101 And yet Fulfillment's pom excludes those logging packages: https://github.com/RentTheRunway/fulfillment/blob/master/fulfillment/pom.xml#L16 

It looks like someone may have ran into problems with those logging packages.. any thoughts?
